### PR TITLE
feat: implement plug agent

### PR DESCRIPTION
**Pull request by 0xAaCE**
# feat: implement plug agent

## Summary

Implemente a PlugAgent extended from HTTPAgent of dfinity library to overwrite call, query and readState method to execute it in extension side

### CODE REVIEW

File: src/utils/agent.ts [hunk 1]
```diff
@@ -21,22 +41,164 @@ const DEFAULT_CREATE_AGENT_ARGS: CreateAgentParamsFixed = {
...
+    identity?: Identity | Promise<Identity>
+  ): Promise<SubmitResponse> => {
+    const metadata = getDomainMetadata();
 
-  const agent = new HttpAgent({
-    identity,
-    host,
-  });
+    const canisterIdStr =
+      typeof canisterId === "string" ? canisterId : canisterId.toString();
```
Comment (tomiir) [inline src/utils/agent.ts]: cant you do `canisterId.toString()` directly? I think `String.toString()` exists

File: src/utils/agent.ts [hunk 2]
```diff
@@ -21,22 +41,164 @@ const DEFAULT_CREATE_AGENT_ARGS: CreateAgentParamsFixed = {
...
+    const effectiveCanisterIdStr =
+      typeof options.effectiveCanisterId === "string"
```
Comment (tomiir) [inline src/utils/agent.ts]: same

File: src/utils/agent.ts [hunk 3]
```diff
@@ -21,22 +41,164 @@ const DEFAULT_CREATE_AGENT_ARGS: CreateAgentParamsFixed = {
   host: DEFAULT_HOST,
 };
 
-export const privateCreateAgent = async ({ publicKey, clientRPC, idls, preApprove = false, whitelist = DEFAULT_CREATE_AGENT_ARGS.whitelist, host = DEFAULT_CREATE_AGENT_ARGS.host }) => {
-  const identity = new PlugIdentity(
-    publicKey,
-    signFactory(clientRPC, idls, preApprove),
-    whitelist
-  );
+const callMethodFactory =
```
Comment (tomiir) [inline src/utils/agent.ts]: there's no way a user could interact with a factory directly, right ?

Comment (tomiir) [inline src/utils/agent.ts]: I'd move factories to a different file and leave the agent here

File: src/utils/bigint.ts [hunk 4]
```diff
@@ -1,4 +1,18 @@
...
+            [key]: recursiveParseBigint(val),
+          }),
+          {}
+        );
+  }
+  if (typeof value === "bigint") {
+    return parseInt(value.toString(), 10);
+  }
+  return value;
+};
```
Comment (tomiir) [inline src/utils/bigint.ts:18]: weren't we using a lib for this?

File: src/Provider/index.ts [hunk 5]
```diff
@@ -180,28 +184,30 @@ export default class Provider implements ProviderInterface {
...
-      metadata,
-      {
-        whitelist: canisterList, host: connectionData?.connection?.host,
-      },
-      this.idls,
-      true
-    );
 
-    const sender = (await agent.getPrincipal()).toString();
+    const sender = (await this.getPrincipal({ asString: true })) as string;
```
Comment (rocky-fleek) [inline src/Provider/index.ts:189]: is there any case where this fails ?

File: src/Provider/index.ts [hunk 6]
```diff
@@ -180,28 +184,30 @@ export default class Provider implements ProviderInterface {
...
     );
 
-    const batchAccepted = await this.clientRPC.call({
+    const batchResponse = await this.clientRPC.call({
       handler: "batchTransactions",
       args: [metadata, signInfo],
     });
 
-    if (!batchAccepted) return false;
+    if (!batchResponse.status) return false;
```
Comment (rocky-fleek) [inline src/Provider/index.ts:200]: isnt status a string?

File: src/utils/agent.ts (hunk 3 above)
Comment (0xAaCE) [reply src/utils/agent.ts]: Actually, that not would be a problem. If someone could interact with the factory and they want to send a call request, they gonna need to use callFactory for example and it gonna display a pop up to the user

File: src/utils/bigint.ts (hunk 4 above)
Comment (0xAaCE) [reply src/utils/bigint.ts:18]: That lib give me a lot of problem actually

File: src/Provider/index.ts (hunk 6 above)
Comment (0xAaCE) [reply src/Provider/index.ts:200]: it should be a boolean

**Review by rocky-fleek — APPROVED** [review]

**Pull request merged**